presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -0,0 +1,197 @@
+        PlanNodeIdAllocator planNodeIdAllocator;
+        VariableAllocator variableAllocator;
+
+        public CteEnumerator(CTEInformationCollector cteInformationCollector, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator)
+        {
+            this.cteInformationCollector = cteInformationCollector;
+            this.planNodeIdAllocator = planNodeIdAllocator;
+            this.variableAllocator = variableAllocator;
+        }
+
+        public PlanNode transformPersistentCtes(PlanNode root)
+        {
+            checkArgument(root.getSources().size() == 1, "expected newChildren to contain 1 node");
+            CteTransformerContext context = new CteTransformerContext();
+            PlanNode transformedCte = SimplePlanRewriter.rewriteWith(new CteConsumerTransformer(planNodeIdAllocator, variableAllocator),
+                    root, context);
+            List<PlanNode> topologicalOrderedList = context.getTopologicalOrdering();
+            if (topologicalOrderedList.isEmpty()) {
+                return root;
+            }
+            SequenceNode sequenceNode = new SequenceNode(root.getSourceLocation(), planNodeIdAllocator.getNextId(), topologicalOrderedList,
+                    transformedCte.getSources().get(0));
+            return root.replaceChildren(Arrays.asList(sequenceNode));
+        }
+    }
+
+    public class CteConsumerTransformer
+            extends SimplePlanRewriter<CteTransformerContext>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        private final VariableAllocator variableAllocator;
+
+        public CteConsumerTransformer(PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        {
+            this.idAllocator = idAllocator;
+            this.variableAllocator = variableAllocator;
+        }
+
+        @Override
+        public PlanNode visitCteReference(CteReferenceNode node, RewriteContext<CteTransformerContext> context)
+        {
+            context.get().addDependency(node.getCteName());
+            context.get().setActiveCte(node.getCteName());
+            // So that dependent CTEs are processed properly
+            PlanNode actualSource = context.rewrite(node.getSource(), context.get());
+            context.get().clearActiveCte();
+            CteProducerNode cteProducerSource = new CteProducerNode(node.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    actualSource,
+                    node.getCteName(),
+                    variableAllocator.newVariable("rows", BIGINT), node.getOutputVariables());
+            context.get().addProducer(node.getCteName(), cteProducerSource);
+            return new CteConsumerNode(node.getSourceLocation(), idAllocator.getNextId(), actualSource.getOutputVariables(), node.getCteName());
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<CteTransformerContext> context)
```

Comment:
Because for apply we also need to rewrite subquery using the same context for correlated queries etc.
eg - [test](https://github.com/prestodb/presto/blob/20f0a66a2118eee1d55a12b369df395fd97ff8ab/presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java#L101) case